File: xls/dslx/type_info.py
# Lint as: python3
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Note: we often access "protected" members on instances of the
# known-same-class, but PyLint flags these as potential errors. We just blanket
# disable instead of annotating every line.
# pylint: disable=protected-access

"""Support for carrying type information from type inferencing."""

from typing import Tuple, Optional, Dict

import dataclasses

from xls.dslx import ast
from xls.dslx import span
from xls.dslx import symbolic_bindings as symbind_mod
from xls.dslx.concrete_type import ConcreteType
from xls.dslx.xls_type_error import TypeInferenceError

ImportedInfo = Tuple[ast.Module, 'TypeInfo']


class TypeMissingError(span.PositionalError):
  """Raised when there is no binding from an AST node to its corresponding type.

  This is useful to raise in order to flag free variables that are dependencies
  for type inference; e.g. functions within a module that invoke other top-level
  functions. The type inference system can catch the error, infer the
  dependency, and re-attempt the deduction of the dependent function.
  """

  def __init__(self, node: ast.AstNode, suffix: str = ''):
    assert isinstance(node, ast.AstNode), repr(node)
    message = 'Missing type for AST node: {node}{suffix}'.format(
        node=node, suffix=' :: ' + suffix if suffix else '')
    # We don't know the real span of the user, we rely on the appropriate caller
    # to catch the error and populate this field properly.
    fake_span = span.Span(span.Pos('<fake>', 0, 0), span.Pos('<fake>', 0, 0))
    super(TypeMissingError, self).__init__(message, fake_span)
    self.node = node
    self.suffix = suffix
    self.user = None


@dataclasses.dataclass
class InvocationData:
  """Parametric instantiation information related to an invocation AST node.

  Instance variables:
    symbolic_bindings_map: Maps from symbolic bindings in the caller to the
      corresponding symbolic bindings in the callee.
    instantiations: Type information that is specialized for a particular
      parametric instantiation of an invocation.
  """
  invocation: ast.Invocation
  symbolic_bindings_map: Dict[
      symbind_mod.SymbolicBindings,
      symbind_mod.SymbolicBindings] = dataclasses.field(
          default_factory=dict)
  instantiations: Dict[symbind_mod.SymbolicBindings,
                       'TypeInfo'] = dataclasses.field(default_factory=dict)

  def update(self, other: 'InvocationData') -> None:
    assert self.invocation == other.invocation
    self.symbolic_bindings_map.update(other.symbolic_bindings_map)
    self.instantiations.update(other.instantiations)


@dataclasses.dataclass
class SliceData:
  node: ast.Slice
  bindings_to_start_width: Dict[symbind_mod.SymbolicBindings,
                                Tuple[int, int]] = dataclasses.field(
                                    default_factory=dict)

  def update(self, other: 'SliceData') -> None:
    assert self.node == other.node
    self.bindings_to_start_width.update(other.bindings_to_start_width)


class TypeInfo:
  """Holds {AstNode: ConcreteType} mapping and other type analysis info.

  Easily "chains" onto an existing mapping of node types when entering a scope
  with parametric bindings; e.g. a new node_to_type mapping is created for
  a parametric function's body after parametric instantiation.

  Also raises a TypeMissingError instead of a KeyError when we encounter a node
  that does not have a type known, so that it can be handled in a more specific
  way versus a KeyError.
  """

  def __init__(self, parent: Optional['TypeInfo'] = None):
    self._dict: Dict[ast.AstNode, ConcreteType] = {}
    self._imports: Dict[ast.Import, ImportedInfo] = {}
    self._name_to_const: Dict[ast.NameDef, ast.Constant] = {}
    self._invocations: Dict[ast.Invocation, InvocationData] = {}
    self._slices: Dict[ast.Slice, SliceData] = {}
    self._parent: Optional['TypeInfo'] = parent

  @property
  def parent(self) -> 'TypeInfo':
    return self._parent

  def update(self, other: 'TypeInfo') -> None:
    """Updates this type information object with the data from 'other'."""
    self._dict.update(other._dict)
    self._imports.update(other._imports)
    # Merge in all the invocation information.
    for invocation, data in other._invocations.items():
      if invocation in self._invocations:
        self._invocations[invocation].update(data)
      else:
        self._invocations[invocation] = data
    # Merge in all the slice information.
    for node, data in other._slices.items():
      if node in self._slices:
        self._slices[node].update(data)
      else:
        self._slices[node] = data

  def _top(self) -> 'TypeInfo':
    """Traverses to the "most parent" TypeInfo."""
    this = self
    while this._parent:
      this = this._parent
    return this

  def add_slice_start_width(
      self, node: ast.Slice,
      symbolic_bindings: symbind_mod.SymbolicBindings,
      start_width: Tuple[int, int]):
    """Notes start/width for a slice operation found during type inference."""
    self._top()._slices.setdefault(node, SliceData(
        node)).bindings_to_start_width[symbolic_bindings] = start_width

  def get_slice_start_width(
      self, node: ast.Slice,
      symbolic_bindings: symbind_mod.SymbolicBindings) -> Tuple[int, int]:
    """Retrieves start/width for slice operation found during type inference."""
    return self._top()._slices[node].bindings_to_start_width[symbolic_bindings]

  def add_invocation_symbolic_bindings(
      self, invocation: ast.Invocation,
      caller: symbind_mod.SymbolicBindings,
      callee: symbind_mod.SymbolicBindings) -> None:
    """Notes caller/callee relation of symbolic bindings at an invocation.

    This is kept from type inferencing time for convenience purposes (so it
    doesn't need to be recalculated anywhere; e.g. in the interpreter).

    Args:
      invocation: The invocation node that (may have) caused parametric
        instantiation.
      caller: The caller's symbolic bindings at the point of invocation.
      callee: The callee's computed symbolic bindings for the invocation.
    """
    self._top()._invocations.setdefault(
        invocation,
        InvocationData(invocation)).symbolic_bindings_map[caller] = callee

  def get_invocation_symbolic_bindings(
      self, invocation: ast.Invocation,
      caller: symbind_mod.SymbolicBindings
  ) -> symbind_mod.SymbolicBindings:
    """Returns callee bindings given caller bindings at an invocation."""
    return self._top()._invocations[invocation].symbolic_bindings_map[caller]

  def add_instantiation(self, invocation: ast.Invocation,
                        caller: symbind_mod.SymbolicBindings,
                        type_info: 'TypeInfo') -> None:
    """Adds derived type info for an "instantiation".

    An "instantiation" is an invocation of a parametric function from some
    caller context (given by the invocation / caller symbolic bindings). These
    have /derived/ type information, where the parametric expressions are
    concretized, and have concrete types corresponding to AST nodes in the
    instantiated parametric function.

    Args:
      invocation: The invocation the type information has been generated for.
      caller: The caller's symbolic bindings that caused this instantiation to
        occur.
      type_info: The type information that has been determined for this
        instantiation.
    """
    self._top()._invocations.setdefault(
        invocation,
        InvocationData(invocation)).instantiations[caller] = type_info

  def has_instantiation(self, invocation: ast.Invocation,
                        caller: symbind_mod.SymbolicBindings) -> bool:
    """Returns if there's type info at invocation with given caller bindings."""
    return caller in self._top()._invocations.get(
        invocation, InvocationData(invocation)).instantiations

  def get_instantiation(
      self, invocation: ast.Invocation,
      caller: symbind_mod.SymbolicBindings) -> 'TypeInfo':
    """Retrieves type info for invocation with given caller bindings."""
    return self._top()._invocations[invocation].instantiations[caller]

  def add_import(self, import_node: ast.Import, info: ImportedInfo) -> None:
    assert import_node not in self._imports, import_node
    self._imports[import_node] = info
    self.update(info[1])

  def note_constant(self, name_def: ast.NameDef, constant: ast.Constant):
    self._name_to_const[name_def] = constant

  def get_const_int(self, name_def: ast.NameDef, user_span: span.Span) -> int:
    if name_def not in self._name_to_const and self.parent:
      return self.parent.get_const_int(name_def, user_span)

    constant = self._name_to_const[name_def]
    value = constant.value
    if isinstance(value, ast.Number):
      return value.get_value_as_int()
    raise TypeInferenceError(
        span=user_span,
        type_=None,
        suffix='Expected to find a constant integral value with the name {};'
        'got: {}'.format(name_def, constant.value))

  def get_imports(self) -> Dict[ast.Import, ImportedInfo]:
    return self._imports if not self.parent else {
        **self._imports,
        **self.parent._imports
    }

  def get_imported(self, import_node: ast.Import) -> ImportedInfo:
    if self.parent:
      if import_node in self._imports:
        return self._imports[import_node]

      return self.parent._imports[import_node]

    return self._imports[import_node]

  def __setitem__(self, k: ast.AstNode, v: ConcreteType) -> None:
    self._dict[k] = v

  def __getitem__(self, k: ast.AstNode) -> ConcreteType:
    """Attempts to resolve AST node 'k' in the node-to-type dictionary.

    Args:
      k: The AST node to resolve to a concrete type.

    Raises:
      TypeMissingError: When the node is not found.

    Returns:
      The previously-determined type of the AST node 'k'.
    """
    assert isinstance(k, ast.AstNode), repr(k)
    try:
      if k in self._dict:
        return self._dict[k]
      if self.parent:
        return self.parent.__getitem__(k)
    except KeyError:
      span_suffix = ' @ {}'.format(k.span) if hasattr(k, 'span') else ''
      raise TypeMissingError(
          k, suffix='resolving type of node{}'.format(span_suffix))
    else:
      span_suffix = ' @ {}'.format(k.span) if hasattr(k, 'span') else ''
      raise TypeMissingError(
          k,
          suffix='resolving type of {} node{}'.format(k.__class__.__name__,
                                                      span_suffix))

  def __contains__(self, k: ast.AstNode) -> bool:
    return (k in self._dict or self.parent.__contains__(k)
            if self.parent else k in self._dict)

File: xls/dslx/type_info.cc
// Copyright 2020 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "xls/dslx/type_info.h"

#include "absl/status/status.h"
#include "xls/common/status/ret_check.h"

namespace xls::dslx {

std::string InvocationData::ToString() const {
  return absl::StrCat("[",
                      absl::StrJoin(symbolic_bindings_map, ", ",
                                    [](std::string* out, const auto& item) {
                                      absl::StrAppendFormat(
                                          out, "%s: %s", item.first.ToString(),
                                          item.second.ToString());
                                    }),
                      "]");
}

// -- class TypeInfoOwner

absl::StatusOr<TypeInfo*> TypeInfoOwner::New(Module* module, TypeInfo* parent) {
  // Note: private constructor so not using make_unique.
  type_infos_.push_back(absl::WrapUnique(new TypeInfo(module, parent)));
  TypeInfo* result = type_infos_.back().get();
  if (parent == nullptr) {
    // Check we only have a single nullptr-parent TypeInfo for a given module.
    XLS_RET_CHECK(!module_to_root_.contains(module))
        << "module " << module->name() << " already has a root TypeInfo";
    XLS_VLOG(5) << "Making root type info for module: " << module->name()
                << " @ " << result;
    module_to_root_[module] = result;
  } else {
    // Check that we don't have parent links that traverse modules.
    XLS_RET_CHECK_EQ(parent->module(), module);
    XLS_VLOG(5) << "Making derived type info for module: " << module->name()
                << " parent: " << parent << " @ " << result;
  }
  return result;
}

absl::StatusOr<TypeInfo*> TypeInfoOwner::GetRootTypeInfo(const Module* module) {
  auto it = module_to_root_.find(module);
  if (it == module_to_root_.end()) {
    return absl::NotFoundError(absl::StrCat(
        "Could not find (root) type information for module: ", module->name()));
  }
  return it->second;
}

// -- class TypeInfo

void TypeInfo::NoteConstExpr(const AstNode* const_expr, InterpValue value) {
  const_exprs_.insert({const_expr, value});
}

absl::StatusOr<InterpValue> TypeInfo::GetConstExpr(
    const AstNode* const_expr) const {
  if (auto it = const_exprs_.find(const_expr); it != const_exprs_.end()) {
    return it->second.value();
  }

  if (parent_ != nullptr) {
    // In a case where a [child] type info is for a parametric function
    // specialization, it won't contain top-level constants, but its parent
    // [transitively] will.
    return parent_->GetConstExpr(const_expr);
  }

  return absl::NotFoundError(absl::StrCat(
      "No constexpr value found for node \"", const_expr->ToString(), "\"."));
}

bool TypeInfo::IsKnownConstExpr(const AstNode* node) {
  if (const_exprs_.contains(node)) {
    return const_exprs_.at(node).has_value();
  }

  if (parent_ != nullptr) {
    return parent_->IsKnownConstExpr(node);
  }

  return false;
}

bool TypeInfo::IsKnownNonConstExpr(const AstNode* node) {
  if (const_exprs_.contains(node)) {
    return !const_exprs_.at(node).has_value();
  }

  if (parent_ != nullptr) {
    return parent_->IsKnownNonConstExpr(node);
  }

  return false;
}

bool TypeInfo::Contains(AstNode* key) const {
  XLS_CHECK_EQ(key->owner(), module_);
  return dict_.contains(key) || (parent_ != nullptr && parent_->Contains(key));
}

std::string TypeInfo::GetImportsDebugString() const {
  return absl::StrFormat(
      "module %s imports:\n  %s", module()->name(),
      absl::StrJoin(imports_, "\n  ", [](std::string* out, const auto& item) {
        absl::StrAppend(out, item.second.module->name());
      }));
}

std::optional<ConcreteType*> TypeInfo::GetItem(const AstNode* key) const {
  XLS_CHECK_EQ(key->owner(), module_)
      << key->owner()->name() << " vs " << module_->name()
      << " key: " << key->ToString();
  auto it = dict_.find(key);
  if (it != dict_.end()) {
    return it->second.get();
  }
  if (parent_ != nullptr) {
    return parent_->GetItem(key);
  }
  return absl::nullopt;
}

absl::StatusOr<ConcreteType*> TypeInfo::GetItemOrError(
    const AstNode* key) const {
  auto maybe_type = GetItem(key);
  if (maybe_type.has_value()) {
    return *maybe_type;
  }

  return absl::NotFoundError(
      absl::StrCat("Could not find concrete type for node: ", key->ToString()));
}

void TypeInfo::AddInvocationCallBindings(const Invocation* call,
                                            SymbolicBindings caller,
                                            SymbolicBindings callee) {
  XLS_CHECK_EQ(call->owner(), module_);
  TypeInfo* top = GetRoot();
  XLS_VLOG(3) << "Type info " << top
              << " adding instantiation call bindings for invocation: "
              << call->ToString() << " @ " << call->span()
              << " caller: " << caller.ToString()
              << " callee: " << callee.ToString();
  auto it = top->invocations_.find(call);
  if (it == top->invocations_.end()) {
    absl::flat_hash_map<SymbolicBindings, SymbolicBindings> symbind_map;
    symbind_map.emplace(std::move(caller), std::move(callee));
    top->invocations_[call] =
        InvocationData{call, std::move(symbind_map)};
    return;
  }
  XLS_VLOG(3) << "Adding to existing invocation data.";
  InvocationData& data = it->second;
  data.symbolic_bindings_map.emplace(std::move(caller), std::move(callee));
}

std::optional<bool> TypeInfo::GetRequiresImplicitToken(
    const Function* f) const {
  XLS_CHECK_EQ(f->owner(), module_) << "function owner: " << f->owner()->name()
                                    << " module: " << module_->name();
  const TypeInfo* root = GetRoot();
  const absl::flat_hash_map<const Function*, bool>& map =
      root->requires_implicit_token_;
  auto it = map.find(f);
  if (it == map.end()) {
    return absl::nullopt;
  }
  bool result = it->second;
  XLS_VLOG(6) << absl::StreamFormat("GetRequiresImplicitToken %p %s::%s => %s",
                                    root, f->owner()->name(), f->identifier(),
                                    (result ? "true" : "false"));
  return result;
}

void TypeInfo::NoteRequiresImplicitToken(const Function* f, bool is_required) {
  TypeInfo* root = GetRoot();
  XLS_VLOG(6) << absl::StreamFormat(
      "NoteRequiresImplicitToken %p: %s::%s => %s", root, f->owner()->name(),
      f->identifier(), is_required ? "true" : "false");
  root->requires_implicit_token_.emplace(f, is_required);
}

std::optional<TypeInfo*> TypeInfo::GetInvocationTypeInfo(
    const Invocation* invocation, const SymbolicBindings& caller) const {
  XLS_CHECK_EQ(invocation->owner(), module_)
      << invocation->owner()->name() << " vs " << module_->name();
  const TypeInfo* top = GetRoot();
  auto it = top->invocations_.find(invocation);
  if (it == top->invocations_.end()) {
    XLS_VLOG(5) << "Could not find instantiation for invocation: "
                << invocation->ToString();
    return absl::nullopt;
  }
  const InvocationData& data = it->second;
  XLS_VLOG(5) << "Invocation " << invocation->ToString()
              << " caller bindings: " << caller
              << " invocation data: " << data.ToString();
  auto it2 = data.instantiations.find(caller);
  if (it2 == data.instantiations.end()) {
    return absl::nullopt;
  }
  return it2->second;
}

absl::StatusOr<TypeInfo*> TypeInfo::GetInvocationTypeInfoOrError(
    const Invocation* invocation, const SymbolicBindings& caller) const {
  auto maybe_ti = GetInvocationTypeInfo(invocation, caller);
  if (maybe_ti.has_value()) {
    return *maybe_ti;
  }

  return absl::NotFoundError(
      absl::StrCat("Could not find child type info with caller bindngs: ",
                   caller.ToString()));
}

void TypeInfo::SetInvocationTypeInfo(const Invocation* invocation,
                                     SymbolicBindings caller,
                                     TypeInfo* type_info) {
  XLS_CHECK_EQ(invocation->owner(), module_);
  TypeInfo* top = GetRoot();
  InvocationData& data = top->invocations_[invocation];
  data.instantiations[caller] = type_info;
}

absl::Status TypeInfo::SetTopLevelProcTypeInfo(const Proc* p, TypeInfo* ti) {
  if (parent_ != nullptr) {
    return absl::InvalidArgumentError(
        "SetTopLevelTypeInfo may only be called on a module top-level type "
        "info.");
  }
  XLS_RET_CHECK_EQ(p->owner(), module_);
  top_level_proc_type_info_[p] = ti;
  return absl::OkStatus();
}

absl::StatusOr<TypeInfo*> TypeInfo::GetTopLevelProcTypeInfo(const Proc* p) {
  if (!top_level_proc_type_info_.contains(p)) {
    return absl::NotFoundError(absl::StrCat(
        "Top-level type info not found for proc \"", p->identifier(), "\"."));
  }
  return top_level_proc_type_info_.at(p);
}

std::optional<const SymbolicBindings*>
TypeInfo::GetInvocationCalleeBindings(const Invocation* invocation,
                                      const SymbolicBindings& caller) const {
  XLS_CHECK_EQ(invocation->owner(), module_)
      << invocation->owner()->name() << " vs " << module_->name();
  const TypeInfo* top = GetRoot();
  XLS_VLOG(3) << absl::StreamFormat(
      "TypeInfo %p getting instantiation symbolic bindings: %p %s @ %s %s", top,
      invocation, invocation->ToString(),
      invocation->span().ToString(), caller.ToString());
  auto it = top->invocations().find(invocation);
  if (it == top->invocations().end()) {
    XLS_VLOG(3) << "Could not find instantiation " << invocation
                << " in top-level type info: " << top;
    return absl::nullopt;
  }
  const InvocationData& data = it->second;
  auto it2 = data.symbolic_bindings_map.find(caller);
  if (it2 == data.symbolic_bindings_map.end()) {
    XLS_VLOG(3)
        << "Could not find caller symbolic bindings in instantiation data: "
        << caller.ToString() << " " << invocation->ToString() << " @ "
        << invocation->span();
    return absl::nullopt;
  }
  const SymbolicBindings* result = &it2->second;
  XLS_VLOG(3) << "Resolved instantiation symbolic bindings for "
              << invocation->ToString() << ": " << result->ToString();
  return result;
}

void TypeInfo::AddSliceStartAndWidth(Slice* node,
                                     const SymbolicBindings& symbolic_bindings,
                                     StartAndWidth start_width) {
  XLS_CHECK_EQ(node->owner(), module_);
  TypeInfo* top = GetRoot();
  auto it = top->slices_.find(node);
  if (it == top->slices_.end()) {
    top->slices_[node] =
        SliceData{node, {{symbolic_bindings, std::move(start_width)}}};
  } else {
    top->slices_[node].bindings_to_start_width.emplace(symbolic_bindings,
                                                       std::move(start_width));
  }
}

std::optional<StartAndWidth> TypeInfo::GetSliceStartAndWidth(
    Slice* node, const SymbolicBindings& symbolic_bindings) const {
  XLS_CHECK_EQ(node->owner(), module_);
  const TypeInfo* top = GetRoot();
  auto it = top->slices_.find(node);
  if (it == top->slices_.end()) {
    return absl::nullopt;
  }
  const SliceData& data = it->second;
  auto it2 = data.bindings_to_start_width.find(symbolic_bindings);
  if (it2 == data.bindings_to_start_width.end()) {
    return absl::nullopt;
  }
  return it2->second;
}

void TypeInfo::AddImport(Import* import, Module* module, TypeInfo* type_info) {
  XLS_CHECK_EQ(import->owner(), module_);
  GetRoot()->imports_[import] = ImportedInfo{module, type_info};
}

std::optional<const ImportedInfo*> TypeInfo::GetImported(
    Import* import) const {
  XLS_CHECK_EQ(import->owner(), module_)
      << "Import node from: " << import->owner()->name() << " vs TypeInfo for "
      << module_->name();
  auto* self = GetRoot();
  auto it = self->imports_.find(import);
  if (it == self->imports_.end()) {
    return absl::nullopt;
  }
  return &it->second;
}

absl::StatusOr<const ImportedInfo*> TypeInfo::GetImportedOrError(
    Import* import) const {
  auto maybe_imported = GetImported(import);
  if (maybe_imported.has_value()) {
    return maybe_imported.value();
  }

  return absl::NotFoundError(
      absl::StrCat("Could not find import for \"", import->ToString(), "\"."));
}

std::optional<TypeInfo*> TypeInfo::GetImportedTypeInfo(Module* m) {
  TypeInfo* root = GetRoot();
  if (root != this) {
    return root->GetImportedTypeInfo(m);
  }
  if (m == module()) {
    return this;
  }
  for (auto& [import, info] : imports_) {
    if (info.module == m) {
      return info.type_info;
    }
  }
  return absl::nullopt;
}

TypeInfo::TypeInfo(Module* module, TypeInfo* parent)
    : module_(module), parent_(parent) {
  XLS_VLOG(6) << "Created type info for module \"" << module_->name() << "\" @ "
              << this << " parent " << parent << " root " << GetRoot();
}

}  // namespace xls::dslx
